From b57e26a7ff3903f66270b9baba404d42b703a2c0 Mon Sep 17 00:00:00 2001
From: Pavan S <psreenivas@habana.ai>
Date: Mon, 04 Mar 2024 11:37:43 +0200
Subject: [PATCH] [SW-176446] gaudi3: add new reset control sequence

This patch adds new reset sequence in soc init which
will be called by arcmgmt in hw init. The new reset
sequence is as per the embedded PRM.

Signed-off-by: Pavan S <psreenivas@habana.ai>
Change-Id: I7bfc18e5fe035fd69782aa868d46835408d46347
---

diff --git a/include/gaudi3/hl_noc_cfg_defs.h b/include/gaudi3/hl_noc_cfg_defs.h
index fa83c38..51a37f6 100644
--- a/include/gaudi3/hl_noc_cfg_defs.h
+++ b/include/gaudi3/hl_noc_cfg_defs.h
@@ -22,5 +22,6 @@
 void gaudi3_init_cache(const struct hl_noc_cfg_info *cfg);
 void gaudi3_init_mc(const struct hl_noc_cfg_info *cfg);
 void gaudi3_init_axiovrd(const struct hl_noc_cfg_info *cfg);
+void gaudi3_rst_ctrl_init(void);
 
 #endif /* NOC_CFG_INCLUDE_GAUDI3_HL_NOC_CFG_DEFS_H_ */
diff --git a/src/gaudi3/rst_ctrl.c b/src/gaudi3/rst_ctrl.c
new file mode 100644
index 0000000..2eb6bd6
--- /dev/null
+++ b/src/gaudi3/rst_ctrl.c
@@ -0,0 +1,195 @@
+// SPDX-License-Identifier: Apache-2.0
+/*
+ * Copyright 2024 HabanaLabs, Ltd.
+ * All Rights Reserved.
+ */
+
+#include <psoc_reset_conf_regs.h>
+#include <hl_platform_defs.h>
+#include <hl_noc_cfg_defs.h>
+#include <hl_fw_if.h>
+
+/* The registers in the array are as per the reset control sequence */
+static const uint32_t rst_ctrl[] = {
+	mmPSOC_RESET_CONF_PSOC_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_ARC_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_ARC_1_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_ARC_2_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_RRTR_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_RRTR_1_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_RRTR_2_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_RRTR_3_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_DRTR_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_NIC_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_NIC_1_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_NIC_2_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_NIC_3_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_NIC_4_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_NIC_5_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_NIC_QMAN_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_NIC_QMAN_1_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_NIC_QMAN_2_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_NIC_QMAN_3_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_NIC_QMAN_4_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_NIC_QMAN_5_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_NIC_PRT_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_NIC_PRT_1_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_NIC_PRT_2_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_NIC_PRT_3_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_NIC_PRT_4_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_NIC_PRT_5_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PDMA_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_SCHED_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_SCHED_1_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_SCHED_2_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_SCHED_3_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_SM_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_SM_1_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_SM_2_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_SM_3_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_NCH_FT_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PCIE_IF_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_NRTR_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_CPU_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PMMU_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_NCH_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_D2D_FT_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_GRTR_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_CS_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_CS_1_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_CS_2_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_CS_3_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PCIE_CTRL_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PCIE_PHY_CFG_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_DIV_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_DIV_1_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_DIV_2_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_DIV_3_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_DIV_4_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_DIV_5_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_HBM_DIV_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_HBM_DIV_1_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_HBM_DIV_2_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_HBM_DIV_3_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PM_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PM_1_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PM_2_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PM_3_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TSTDVS_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TSTDVS_1_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TSTDVS_2_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TSTDVS_3_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TSTDVS_IF_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TSTDVS_IF_1_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TSTDVS_IF_2_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TSTDVS_IF_3_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_DPHY_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_DPHY_1_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_DPHY_2_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_DPHY_3_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_DPHY_4_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_DPHY_5_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_D2D_MAC_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_D2D_MAC_1_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PLL_PSOC_ARC_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PLL_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PLL_1_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PLL_2_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PLL_3_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PLL_4_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PLL_5_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PLL_6_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PLL_7_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PLL_8_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PLL_9_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PLL_10_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PLL_11_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PLL_12_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PLL_13_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PLL_14_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PLL_15_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PLL_16_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PLL_17_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PLL_18_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PLL_19_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PLL_20_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PLL_21_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PLL_22_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_PLL_23_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_MME_EUS_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_MME_EUS_1_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_MME_EUS_2_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_MME_EUS_3_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_MSS_CLS_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_MSS_CLS_1_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_MSS_CLS_2_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_MSS_CLS_3_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_1_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_2_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_3_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_4_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_5_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_6_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_7_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_8_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_9_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_10_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_11_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_12_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_13_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_14_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_15_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_16_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_17_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_18_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_19_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_20_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_21_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_22_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_23_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_24_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_25_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_26_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_27_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_28_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_29_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_30_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_31_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_ECLS_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_TPC_ECLS_1_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_STLB_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_STLB_1_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_STLB_2_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_STLB_3_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_EDMA_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_EDMA_1_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_ROTATOR_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_ROTATOR_1_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_ROTATOR_2_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_ROTATOR_3_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_VIDEO_DEC_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_VIDEO_DEC_1_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_VIDEO_DEC_2_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_VIDEO_DEC_3_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_VIDEO_DEC_4_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_VIDEO_DEC_5_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_VIDEO_DEC_6_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_VIDEO_DEC_7_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_HBM_MC_0_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_HBM_MC_1_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_HBM_MC_2_CLK_RST_CTRL,
+	mmPSOC_RESET_CONF_HBM_MC_3_CLK_RST_CTRL
+};
+
+void gaudi3_rst_ctrl_init(void)
+{
+	uint32_t iter;
+	uint8_t sku = hl_get_sku_type();
+
+	for (iter = 0; iter < ARRAY_SIZE(rst_ctrl); iter++) {
+		CFG_WREG32(D0_PSOC_RESET_CONF_REG(rst_ctrl[iter]), iter);
+		if (sku == HL_DUAL_DIE_SKU)
+			CFG_WREG32(D1_PSOC_RESET_CONF_REG(rst_ctrl[iter]), iter);
+	}
+}
